src/Catalog/Persistence/AzureStorage.cs

```diff
@@ -125,8 +124,43 @@ private StorageListItem GetStorageListItem(IListBlobItem listBlobItem)
                     Trace.WriteLine(String.Format("Saved uncompressed blob {0} to container {1}", blob.Uri.ToString(), _directory.Container.Name));
                 }
             }
+            TryTakeBlobSnapshot(blob);
         }
 
+        /// <summary>
+        /// Take one snapshot only if there is not any snapshot for the specific blob
+        /// This will prevent the blob to be deleted by a not intended delete action
+        /// </summary>
+        /// <param name="blob"></param>
+        /// <returns></returns>
+        private bool TryTakeBlobSnapshot(CloudBlockBlob blob)
+        {
+            if (blob == null )
+            {
+                //no action
+                return false;
+            }
+            Stopwatch sw = new Stopwatch();
+            sw.Start();
+            try
+            {
+                var allSnapshots = blob.Container.ListBlobs(prefix: blob.Name, useFlatBlobListing: true, blobListingDetails: BlobListingDetails.Snapshots);
```

Comment:
nit: please wrap long lines (120 is good b/c GitHub source viewing is most natural)